contracts/src/AgentExecutor.sol

```diff
@@ -36,4 +50,25 @@ contract AgentExecutor {
     function _transferToken(address token, address recipient, uint128 amount) internal {
         IERC20(token).safeTransfer(recipient, amount);
     }
+
+    /// @dev Register native asset from polkadto as ERC20 `token`.
+    function _registerToken(bytes32 agentID, bytes32 tokenID, string memory name, string memory symbol, uint8 decimals)
+        internal
+    {
+        IERC20 token = new ERC20(name, symbol, decimals);
+        Gateway(msg.sender).registerForeignToken(tokenID, address(token), agentID);
+    }
+
+    /// @dev Mint ERC20 token to `recipient`.
+    function _mintToken(bytes32 tokenID, address recipient, uint256 amount) internal {
+        address token = Gateway(msg.sender).tokenAddressOf(tokenID);
```

Comment:
`token` can be resolved in Gateway and passed to AgentExecutor as a parameter. This will brake the dependency on `msg.sender`.

Reply:
https://github.com/Snowfork/snowbridge/pull/1155/commits/28f513d669b7bf4e97ac60e719147694599504c6